File: examples/python/textual/notes_showcase/src/notes_showcase/viewmodels/notebooks_root_vm.py
"""NotebooksRootVM — root of the notebooks tree.

VMx-API adaptation (cross-reference with the C# §3.a flavor):
:class:`vmx.HierarchicalVM` materialises its children eagerly from a factory at
construct time and is awkward to mutate dynamically (it supports
``add_child``/``remove_child`` but has no first-class "current selection"
surface the plan assumed). Instead we own a flat
:class:`~vmx.ObservableList` of :class:`NotebookVM` instances, exposing
``roots`` / ``children_of`` walks for the view, and publish a
:class:`vmx.TreeStructureChangedMessage` on every add/remove so subscribers
observe structural changes identically to the HierarchicalVM contract.
"""

from __future__ import annotations

import asyncio
import dataclasses
import uuid

from vmx import (
    ComponentVM,
    ConstructionStatus,
    INewCreatable,
    IReconstructable,
    MessageHub,
    ObservableList,
    PropertyChangedMessage,
    RelayCommand,
    RxDispatcher,
    TreeStructureChange,
    TreeStructureChangedMessage,
)
from vmx.messages.protocols import Message
from vmx.notifications import INotificationHub, Notification, NotificationType
from vmx.services.dispatcher import Dispatcher

from notes_showcase.models.note_repository import INoteRepository
from notes_showcase.models.notebook_model import NotebookModel
from notes_showcase.viewmodels.notebook_vm import NotebookVM


class NotebooksRootVM(ComponentVM, INewCreatable, IReconstructable):
    """Root VM owning the flat collection of notebook VMs."""

    def __init__(
        self,
        *,
        name: str,
        hint: str,
        hub: MessageHub[Message],
        dispatcher: Dispatcher,
        repository: INoteRepository,
        notification_hub: INotificationHub | None = None,
    ) -> None:
        super().__init__(name=name, hint=hint, hub=hub, dispatcher=dispatcher)
        self._repo = repository
        self._notification_hub = notification_hub
        self._notebooks: ObservableList[NotebookVM] = ObservableList()
        self._current: NotebookVM | None = None

        self._add_notebook_command = (
            RelayCommand.builder()
            .predicate(self.can_create_new)
            .task(self.create_new)
            .build()
        )

    # ── Convenience hub accessor ───────────────────────────────────────────
    @property
    def hub(self) -> MessageHub[Message]:
        return self._hub

    # ── Collection projections ─────────────────────────────────────────────
    @property
    def all(self) -> ObservableList[NotebookVM]:
        """All notebook VMs in load order."""
        return self._notebooks

    @property
    def roots(self) -> list[NotebookVM]:
        """Top-level notebook VMs (``parent_id is None``)."""
        return [nb for nb in self._notebooks if nb.model.parent_id is None]

    def children_of(self, parent: NotebookVM) -> list[NotebookVM]:
        """Children of *parent* by ``parent_id`` walk."""
        return [nb for nb in self._notebooks if nb.model.parent_id == parent.model.id]

    def walk(self) -> list[NotebookVM]:
        """All notebook VMs (parents-before-children, repo order)."""
        return list(self._notebooks)

    # ── current (two-way bindable) ─────────────────────────────────────────
    @property
    def current(self) -> NotebookVM | None:
        return self._current

    @current.setter
    def current(self, value: NotebookVM | None) -> None:
        if self._current is value:
            return
        self._current = value
        self._hub.send(PropertyChangedMessage.create(self, self._name, "current"))
        self._raise_property_changed("current")

    # ── INewCreatable ──────────────────────────────────────────────────────
    def can_create_new(self) -> bool:
        return self._status == ConstructionStatus.CONSTRUCTED

    def create_new(self) -> None:
        """Fire-and-forget schedule of a default "New Notebook" add."""
        # Schedule via asyncio if a loop is running; otherwise no-op (the
        # async add_notebook is the canonical entry-point).
        try:
            loop = asyncio.get_running_loop()
            loop.create_task(self.add_notebook(parent_id=None, name="New Notebook"))
        except RuntimeError:
            asyncio.run(self.add_notebook(parent_id=None, name="New Notebook"))

    @property
    def add_notebook_command(self) -> RelayCommand:
        return self._add_notebook_command

    # ── Async mutation ─────────────────────────────────────────────────────
    async def add_notebook(self, parent_id: str | None, name: str) -> NotebookVM:
        """Persist a new notebook via the repo and publish TreeStructureChanged.

        Returns the newly created :class:`NotebookVM`.
        """
        nb_id = f"nb-{uuid.uuid4().hex[:6]}"
        model = NotebookModel(id=nb_id, name=name, parent_id=parent_id)
        await self._repo.add_notebook(model)
        vm = (
            NotebookVM.builder()
            .name(f"nb:{nb_id}")
            .services(self.hub, self._dispatcher)
            .model(model)
            .children_getter(self.children_of)
            .build()
        )
        vm.construct()
        index = self._notebooks.count
        self._notebooks.append(vm)
        self._hub.send(
            TreeStructureChangedMessage(
                sender=self,
                sender_name=self._name,
                change=TreeStructureChange.ADDED,
                affected=vm,
                index=index,
            )
        )
        # Spec §6.2: publish a "Notebook added" notification (cross-flavor
        # parity with the C# and TypeScript flavors). No-op when no
        # notification hub is wired.
        if self._notification_hub is not None:
            self._notification_hub.post(
                Notification(
                    NotificationType.NOTIFICATION,
                    f'Notebook added: “{name}”',
                )
            )
        return vm

    async def populate(self) -> None:
        """Replace the inner collection with fresh data from the repo.

        Disposes any previously-held VMs first. Called by
        :meth:`WorkspaceVM.construct_async` during workspace startup.
        """
        notebooks, _ = await self._repo.load_all()
        # Dispose previous children.
        for prev in list(self._notebooks):
            prev.dispose()
        # Replace via a single Reset emission rather than per-item churn.
        with self._notebooks.batch_update():
            while self._notebooks.count > 0:
                self._notebooks.remove_at(0)
            for nb in notebooks:
                vm = (
                    NotebookVM.builder()
                    .name(f"nb:{nb.id}")
                    .services(self.hub, self._dispatcher)
                    .model(nb)
                    .children_getter(self.children_of)
                    .build()
                )
                vm.construct()
                self._notebooks.append(vm)
        self._current = None
        # Re-broadcast a tree-structure reset so subscribers refresh.
        self._hub.send(
            TreeStructureChangedMessage(
                sender=self,
                sender_name=self._name,
                change=TreeStructureChange.ADDED,
                affected=self,
                index=-1,
            )
        )

    # ── Lifecycle overrides ────────────────────────────────────────────────
    def _on_destruct(self) -> None:
        for nb in list(self._notebooks):
            nb.destruct()
        super()._on_destruct()

    def _on_dispose(self) -> None:
        for nb in list(self._notebooks):
            nb.dispose()
        self._add_notebook_command.dispose()
        super()._on_dispose()

    # ── Builder entry-point ────────────────────────────────────────────────
    @staticmethod
    def builder() -> NotebooksRootVMBuilder:  # type: ignore[override]
        # Narrows ComponentVM.builder() to the showcase NotebooksRootVMBuilder.
        return NotebooksRootVMBuilder()


@dataclasses.dataclass(frozen=True, slots=True)
class NotebooksRootVMBuilder:
    """Immutable fluent builder for :class:`NotebooksRootVM`."""

    _name: str | None = None
    _hint: str = ""
    _hub: MessageHub[Message] | None = None
    _dispatcher: Dispatcher | None = None
    _repo: INoteRepository | None = None
    _notification_hub: INotificationHub | None = None

    def name(self, value: str) -> NotebooksRootVMBuilder:
        return dataclasses.replace(self, _name=value)

    def hint(self, value: str) -> NotebooksRootVMBuilder:
        return dataclasses.replace(self, _hint=value)

    def services(self, hub: MessageHub[Message], dispatcher: Dispatcher) -> NotebooksRootVMBuilder:
        return dataclasses.replace(self, _hub=hub, _dispatcher=dispatcher)

    def repository(self, repo: INoteRepository) -> NotebooksRootVMBuilder:
        return dataclasses.replace(self, _repo=repo)

    def notification_hub(self, nh: INotificationHub) -> NotebooksRootVMBuilder:
        return dataclasses.replace(self, _notification_hub=nh)

    def build(self) -> NotebooksRootVM:
        if self._name is None:
            raise ValueError("name is required")
        if self._repo is None:
            raise ValueError("repository is required")
        hub = self._hub if self._hub is not None else MessageHub[Message]()
        dispatcher = self._dispatcher if self._dispatcher is not None else RxDispatcher.immediate()
        return NotebooksRootVM(
            name=self._name,
            hint=self._hint,
            hub=hub,
            dispatcher=dispatcher,
            repository=self._repo,
            notification_hub=self._notification_hub,
        )

File: examples/swift/notes-showcase/Sources/NotesShowcaseCore/ViewModels/NotebooksRootVM.swift
//
// NotebooksRootVM — root of the notebooks tree.
//
// Ports examples/csharp/avalonia/NotesShowcase/ViewModels/NotebooksRootVM.cs.
// See task-6-brief.md.
//
// VMx-API adaptation: owns a flat [NotebookVM] instead of a HierarchicalVM
// (same rationale as the C# port — flat mutation + ChildrenGetter closure is
// a cleaner fit for a freely-mutated parent-id collection). Publishes a
// library `TreeStructureChangedMessage` on structural mutations so
// subscribers observe changes identically to HierarchicalVM.
//
// Cross-module subclassing enabled by ADR-0066: `hub`, `dispatcher`, and
// `_notifyPropertyChanged` are `public` on `ComponentVMBase`.
//
import Foundation
import Combine
import VMx

/// Root view-model for the notebooks tree.
///
/// Conforms to `NewCreatable` (create-a-blank-notebook affordance);
/// `Reconstructable` is satisfied automatically via the library extension
/// `extension ComponentVMBase: Constructable, Destructable, Reconstructable {}`.
public final class NotebooksRootVM: ComponentVMBase, NewCreatable {

    // ── Private state ──────────────────────────────────────────────────────

    private let _repo: any NoteRepository
    private let _notificationHub: NotificationHubProtocol?
    private var _all: [NotebookVM] = []
    private var _current: NotebookVM?

    // ── Commands ───────────────────────────────────────────────────────────

    /// Convenience command wired to `canCreateNew()` / `createNew()`.
    public private(set) var addNotebookCommand: AsyncRelayCommand

    // ── Public tree surface ────────────────────────────────────────────────

    /// Flat ordered list of all notebook VMs (all levels, insertion order).
    public var all: [NotebookVM] { _all }

    /// Root notebooks (no parent).
    public var roots: [NotebookVM] { _all.filter { $0.model.parentId == nil } }

    /// All notebooks in insertion order (parents before children, repo order).
    public func walk() -> [NotebookVM] { _all }

    /// Children of `parent` by `parentId` match.
    public func childrenOf(_ parent: NotebookVM) -> [NotebookVM] {
        _all.filter { $0.model.parentId == parent.model.id }
    }

    // ── Current selection ──────────────────────────────────────────────────

    /// Currently selected notebook (reference-guarded setter).
    ///
    /// Setting the same reference is a no-op. Emits a
    /// `PropertyChangedMessage` on the hub and fires `_notifyPropertyChanged`
    /// on change.
    public var current: NotebookVM? {
        get { _current }
        set {
            guard _current !== newValue else { return }
            _current = newValue
            _notifyPropertyChanged("current")
        }
    }

    // ── NewCreatable ───────────────────────────────────────────────────────

    public func canCreateNew() -> Bool { isConstructed }

    /// Synchronous capability entry point routed through the async command.
    public func createNew() {
        addNotebookCommand.execute()
    }

    // ── Async operations ───────────────────────────────────────────────────

    /// Persists a new notebook via the repository, constructs its VM,
    /// appends it to `all`, and publishes a `TreeStructureChangedMessage`.
    /// Optionally posts a "Notebook added" notification when a
    /// `notificationHub` is wired.
    ///
    /// Async emit of structural changes is foreground-marshalled via the
    /// injected dispatcher (mirrors C# `_dispatcher.Foreground.Schedule`).
    public func addNotebook(parentId: String?, name notebookName: String) async throws {
        guard status != .disposed, !Task.isCancelled else { return }
        let uuid = UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
        let id = "nb-\(uuid.prefix(5))"
        let model = NotebookModel(id: id, name: notebookName, parentId: parentId)
        try await _repo.addNotebook(model)
        guard status != .disposed, !Task.isCancelled else { return }

        let added = await runOnForeground { [weak self] in
            guard let self, self.status != .disposed else {
                return false
            }
            let vm = try! NotebookVM.builder()
                .name("nb:\(id)")
                .model(model)
                .services(hub: self.hub, dispatcher: self.dispatcher)
                .childrenGetter({ [weak self] parent in self?.childrenOf(parent) ?? [] })
                .build()
            try? vm.construct()
            self._all.append(vm)

            if let pid = parentId {
                self._all.first(where: { $0.model.id == pid })?
                    .notifyChildrenChanged()
            } else {
                self._notifyPropertyChanged("roots")
            }

            self.hub.send(TreeStructureChangedMessage(
                sender: self,
                senderName: self.name,
                change: .added,
                affected: vm,
                index: self._all.count - 1
            ))
            return true
        }
        guard added, !Task.isCancelled else { return }

        if let notificationHub = _notificationHub {
            publishNotification(VMx.Notification(
                type: .notification,
                message: "Notebook added: \u{201C}\(notebookName)\u{201D}"
            ), to: notificationHub)
        }
    }

    /// Loads all notebooks from the repository, replaces the flat list, and
    /// constructs each child VM. Intended to be called during workspace
    /// async construction.
    public func populate() async throws {
        guard status != .disposed, !Task.isCancelled else { return }
        let (notebooks, _) = try await _repo.loadAll()
        guard status != .disposed, !Task.isCancelled else { return }

        await runOnForeground { [weak self] in
            guard let self, self.status != .disposed else {
                return
            }

            for nb in self._all { nb.dispose() }
            self._all.removeAll()
            self._current = nil
            self._notifyPropertyChanged("current")

            for nb in notebooks {
                let vm = try! NotebookVM.builder()
                    .name("nb:\(nb.id)")
                    .model(nb)
                    .services(hub: self.hub, dispatcher: self.dispatcher)
                    .childrenGetter({ [weak self] parent in self?.childrenOf(parent) ?? [] })
                    .build()
                try? vm.construct()
                self._all.append(vm)
            }

            self.hub.send(TreeStructureChangedMessage(
                sender: self,
                senderName: self.name,
                change: .added,
                affected: self,
                index: -1
            ))
            self._notifyPropertyChanged("roots")
        }
    }

    // ── Init ───────────────────────────────────────────────────────────────

    private init(
        name: String,
        hint: String,
        hub: MessageHubProtocol,
        dispatcher: Dispatcher,
        repo: any NoteRepository,
        notificationHub: NotificationHubProtocol?
    ) {
        _repo = repo
        _notificationHub = notificationHub

        // Phase 1: placeholder command (required before super.init).
        addNotebookCommand = AsyncRelayCommand(
            body: nil,
            predicate: nil,
            triggers: [],
            throwOnCancel: false
        )

        super.init(name: name, hint: hint, hub: hub, dispatcher: dispatcher)

        // Phase 2: rewire with real self-capturing closures.
        addNotebookCommand = AsyncRelayCommand.builder()
            .predicate({ [weak self] in self?.canCreateNew() ?? false })
            .task({ [weak self] in
                guard let self else { return }
                try await self.addNotebook(parentId: nil, name: "New Notebook")
            })
            .build()
    }

    // ── Lifecycle overrides ────────────────────────────────────────────────

    public override func _onDestruct() throws {
        for nb in _all { try? nb.destruct() }
        try super._onDestruct()
    }

    public override func _onDispose() {
        for nb in _all { nb.dispose() }
        addNotebookCommand.dispose()
        super._onDispose()
    }

    // ── Builder ────────────────────────────────────────────────────────────

    /// Returns a new empty builder.
    public static func builder() -> NotebooksRootVMBuilder {
        NotebooksRootVMBuilder()
    }

    /// Immutable fluent builder for `NotebooksRootVM` (spec ch. 10).
    ///
    /// Required: `name(_:)`, `services(hub:dispatcher:)`, `repository(_:)`.
    /// Optional: `hint(_:)`, `notificationHub(_:)`.
    public struct NotebooksRootVMBuilder {
        private var _name: String?
        private var _hint: String = ""
        private var _hub: MessageHubProtocol?
        private var _dispatcher: Dispatcher?
        private var _repo: (any NoteRepository)?
        private var _notificationHub: NotificationHubProtocol?

        fileprivate init() {}

        public func name(_ value: String) -> NotebooksRootVMBuilder {
            var c = self; c._name = value; return c
        }
        public func hint(_ value: String) -> NotebooksRootVMBuilder {
            var c = self; c._hint = value; return c
        }
        public func services(hub: MessageHubProtocol, dispatcher: Dispatcher) -> NotebooksRootVMBuilder {
            var c = self; c._hub = hub; c._dispatcher = dispatcher; return c
        }
        public func repository(_ repo: any NoteRepository) -> NotebooksRootVMBuilder {
            var c = self; c._repo = repo; return c
        }
        public func notificationHub(_ hub: NotificationHubProtocol) -> NotebooksRootVMBuilder {
            var c = self; c._notificationHub = hub; return c
        }

        /// Validates required fields and constructs a `NotebooksRootVM`.
        ///
        /// - Throws: `BuilderValidationError` if `name`, `services`, or
        ///   `repository` are missing.
        public func build() throws -> NotebooksRootVM {
            guard let name = _name else { throw BuilderValidationError(missingField: "name") }
            guard let hub = _hub else { throw BuilderValidationError(missingField: "hub") }
            guard let dispatcher = _dispatcher else { throw BuilderValidationError(missingField: "dispatcher") }
            guard let repo = _repo else { throw BuilderValidationError(missingField: "repository") }
            return NotebooksRootVM(
                name: name, hint: _hint,
                hub: hub, dispatcher: dispatcher,
                repo: repo, notificationHub: _notificationHub
            )
        }
    }
}
